src/qibolab/_core/instruments/qm/config/pulses.py

```diff
@@ -118,30 +118,29 @@ def from_dc_pulse(cls, pulse: Pulse):
 
 def integration_weights(element: str, readout_len: int, kernel=None, angle: float = 0):
     """Create integration weights section for QM config."""
-    cos, sin = np.cos(angle), np.sin(angle)
-    if kernel is None:
 
-        def convert(x):
-            return [(x, readout_len)]
-    else:
-        cos = kernel * cos
-        sin = kernel * sin
+    def convert(x: list[tuple], minus: bool = False):
+        return [(-i[0] if minus else i[0], i[1]) for i in x]
```

Comment:
Since now this is just a pure function, and does not depend on the environment (it is not a closure), I'd suggest moving it top-level.

You can prepend `_` to the name, and possibly make it more specific (`_integration_length_convert`), or anything like that.

Reply:
@andrea-pasquale you partially misunderstood my comment: the main goal was to lift the function outside `integration_weights`, to avoid accidental capture of local variables (which is not happening like this, but it will never happen if it's not part of the scope).

The name change was accidental, just to somehow preserve the scope (and there was a lapsus - it should have been integration **weights**, not *length*... my bad...)

Reply:
Indeed I misunderstood, I have moved the function outside the function scope in https://github.com/qiboteam/qibolab/pull/1155/commits/c9f1eaf4064a97f616c005a299e403e97465d683

Reply:
Thanks!